### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by illright**
* Make each node start and stop its sources
* Refactor the base classes to create "capability" abstract mixins

### CODE REVIEW

**Review by alkaitagi — APPROVED** [review]

File: vpt/capabilities/input_capable.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+from typing import List
+
+from .output_capable import OutputCapable, Initiatable
+
+
+class InputCapable(Initiatable, ABC):
+    '''Base class for input-capable nodes.
+       Should be used for processors and sinks.'''
+    sources: List[OutputCapable]
+
```
Comment (aabounegm) [inline vpt/capabilities/input_capable.py:12]: Shouldn't there be a constructor (that is also an `@abstractmethod`) here that accepts a list of sources and sets `self.stopped = True` (instead of writing it in every subclass - it is better to write `super().__init__()` anyway)?

File: vpt/processors/engagement_estimator/__init__.py [hunk 2]
```diff
@@ -1,6 +0,0 @@
-'''The module responsible for estimating the engagement of the user.'''
```
Comment (aabounegm) [inline vpt/processors/engagement_estimator/__init__.py:1]: Why did you delete this? Wasn't it supposed to be implemented?

File: vpt/processors/mouse_compressor/__init__.py [hunk 3]
```diff
@@ -1,6 +0,0 @@
-'''Compress mouse events.'''
```
Comment (aabounegm) [inline vpt/processors/mouse_compressor/__init__.py:1]: And this as well. It was already implemented in another branch/PR

File: vpt/processors/base.py [hunk 4]
```diff
@@ -1,17 +1,11 @@
 '''The base interfaces for the processor nodes.'''
+from abc import ABC
 from typing import TypeVar
-from abc import abstractmethod, ABC
 
-from rx import Observable
-
-from vpt.sinks.base import SinkBase
-from vpt.sources.base import SourceBase
+from ..capabilities import InputCapable, OutputCapable
```
Comment (aabounegm) [inline vpt/processors/base.py:5]: Prefer module/absolute imports over `..` ones
```suggestion
from vpt.capabilities import InputCapable, OutputCapable
```

File: vpt/sinks/__init__.py [hunk 5]
```diff
@@ -2,3 +2,9 @@
 from .file_store import FileStore
 from .video_display import VideoDisplay
 from .sqlite_store import SQLiteStore
+
+__all__ = (
+    'FileStore',
+    'VideoDisplay',
+    'SQLiteStore',
+)
```
Comment (aabounegm) [inline vpt/sinks/__init__.py:10]: Why do we need this? There aren't any other symbols in this file anyways

File: vpt/sinks/video_display/__init__.py [hunk 6]
```diff
@@ -23,19 +22,19 @@ def __init__(self, video_frame_source: SourceBase[VideoFrame], duration=-1):
             -1 means the window will never close. Given in seconds.
         """
         self.stopped = False
```
Comment (aabounegm) [inline vpt/sinks/video_display/__init__.py]: So it starts working immediately on initialization?
This will block the start of its sources, since `super().start()` checks if it is stopped, and if so doesn't start any source

File: vpt/cli/check.py [hunk 7]
```diff
@@ -14,33 +14,29 @@ def check():
     print('Checking the devices for 5s...')
 
     # Create capture nodes
+    audio_device = sd.query_devices(kind='input', device=args['audio'])
```
Comment (aabounegm) [inline vpt/cli/check.py:17]: I think the `kind` argument is ignored when `device` is passed

File: vpt/sources/device_audio_source/__init__.py [hunk 8]
```diff
@@ -12,43 +12,55 @@
...
+           recorded with `sounddevice`.'''
         eps = 1e-4
-        for sample_idx in range(len(chunk)):
-            if np.any(chunk[sample_idx] > eps):
+        idx = None
+        for idx, sample in enumerate(chunk):
+            if np.any(sample > eps):
                 break
-        return chunk[int(sample_idx * 1.1):, :]
+        return chunk[int(idx * 1.1):, :]
```
Comment (aabounegm) [inline vpt/sources/device_audio_source/__init__.py:54]: `TypeError: unsupported operand type(s) for *: 'NoneType' and 'int'`

File: vpt/capabilities/initiatable.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+       Should be used for sources and processors.'''
+    stopped: bool
+
+    @abstractmethod
+    def start(self):
+        '''Start the node's data stream.'''
+
+    @abstractmethod
+    def stop(self):
+        '''Start the node's data stream.'''
```
Comment (aabounegm) [inline vpt/capabilities/initiatable.py:16]: How about the default implementations of `stopped = True` and `stopped = False` for cases when they suffice?

**Review by kuredoro — APPROVED** [review]

File: vpt/processors/engagement_estimator/__init__.py (hunk 2 above)
Comment (illright) [reply vpt/processors/engagement_estimator/__init__.py:1]: We will bring it in later

File: vpt/sinks/__init__.py (hunk 5 above)
Comment (illright) [reply vpt/sinks/__init__.py:10]: It silences PyLint's warnings about unused variables

**Review by IamMaxim — APPROVED** [review]

File: vpt/cli/check.py (hunk 7 above)
Comment (illright) [reply vpt/cli/check.py:17]: Well, whatever) it worked alright when I tried it

File: vpt/capabilities/initiatable.py (hunk 9 above)
Comment (illright) [reply vpt/capabilities/initiatable.py:16]: They never do

File: vpt/processors/base.py (hunk 4 above)
Comment (illright) [reply vpt/processors/base.py:5]: mmmake me

File: vpt/capabilities/initiatable.py (hunk 9 above)
Comment (aabounegm) [reply vpt/capabilities/initiatable.py:16]: They do in most of the sources. This is already the implementation there

File: vpt/capabilities/input_capable.py (hunk 1 above)
Comment (illright) [reply vpt/capabilities/input_capable.py:12]: Moving the list of sources thing into a common place would be a mess. As for `stopped = True`, I agree, but can't be bothered lmao

**Review by aabounegm — APPROVED** [review]
